src/System.Linq/tests/ZipTests.cs

```diff
@@ -0,0 +1,393 @@
+        }
+        
+        [Fact]
+        public void FuncIsNull()
+        {
+            IEnumerable<int> first = new int[] { 1, 2, 3 };
+            IEnumerable<int> second = new int[] { 2, 4, 6 };
+            Func<int, int, int> func = null;
+            
+            Assert.Throws<ArgumentNullException>(() => first.Zip(second, func));
+        }
+
+        private class MyIEnum<T> : IEnumerable<T>
+        {
+            public IEnumerable<T> _data;
+
+            public MyIEnum(IEnumerable<T> source)
+            {
+                _data = source;
+            }
+
+            public IEnumerator<T> GetEnumerator()
+            {
+                foreach (var datum in _data)
+                {
+                    if (datum.Equals(2)) throw new Exception();
+                    yield return datum;
+                }
+            }
+
+            IEnumerator IEnumerable.GetEnumerator()
+            {
+                return this.GetEnumerator();
+            }
+        }
+
+        [Fact]
+        public void ExceptionThrownFromFirstGetEnumerator()
+        {
+            MyIEnum<int> first = new MyIEnum<int>(new int[] { 1, 3, 3 });
+            IEnumerable<int> second = new int[] { 2, 4, 6 };
+            Func<int, int, int> func = (x, y) => x + y;
+            IEnumerable<int> expected = new int[] { 3, 7, 9 };
+            
+            Assert.Equal(expected, first.Zip(second, func));
+
+            first = new MyIEnum<int>(new int[] { 1, 2, 3 });
+            
+            Assert.ThrowsAny<Exception>(() => first.Zip(second, func).ToList());
+        }
+
+        [Fact]
+        public void ExceptionThrownFromSecondGetEnumerator()
+        {
+            MyIEnum<int> second = new MyIEnum<int>(new int[] { 1, 3, 3 });
+            IEnumerable<int> first = new int[] { 2, 4, 6 };
+            Func<int, int, int> func = (x, y) => x + y;
+            IEnumerable<int> expected = new int[] { 3, 7, 9 };
+
+            Assert.Equal(expected, first.Zip(second, func));
+
+            second = new MyIEnum<int>(new int[] { 1, 2, 3 });
+
+            Assert.ThrowsAny<Exception>(() => first.Zip(second, func).ToList());
+        }
+
+        [Fact]
+        public void FirstAndSecondEmpty()
+        {
+            IEnumerable<int> first = new int[] { };
+            IEnumerable<int> second = new int[] { };
+            Func<int, int, int> func = (x, y) => x + y;
+            IEnumerable<int> expected = new int[] { };
+
+            Assert.Equal(expected, first.Zip(second, func));
+        }
+
+
+        [Fact]
+        public void FirstEmptySecondSingle()
+        {
+            IEnumerable<int> first = new int[] { };
+            IEnumerable<int> second = new int[] { 2 };
+            Func<int, int, int> func = (x, y) => x + y;
+            IEnumerable<int> expected = new int[] { };
+
+            Assert.Equal(expected, first.Zip(second, func));
+        }
```

Comment:
My personal preference is to just put the expected value in directly, but I think there's a value in fidelity to the original when it doesn't do anything either very unidiomatic for XUnit and isn't doing a lot to reinvent XUnit's wheels.